docs: typescript login (#7621)
zitadel/zitadel: docs/docs/guides/integrate/login-ui/typescript-repo.mdx

To replace our current golang built login UI and showcase the use of our new resource, session and OIDC APIs, we've created the [Typescript Repository](https://github.com/zitadel/typescript).
The typescript repository contains all TypeScript and JavaScript packages and applications you need to create your own Login UI using the ZITADEL APIs.

## Included Apps And Packages

- **[login](./typescript-repo#new-login-ui)**: The future login UI used by ZITADEL Cloud, powered by Next.js
- `@zitadel/server`: core components for establishing node client connection, grpc stub
- `@zitadel/client`: core components for establishing web client connection, grpc stub
- `@zitadel/react`: shared React utilities and components built with Tailwind CSS
- `@zitadel/next`: shared Next.js utilities
- `@zitadel/tsconfig`: shared `tsconfig.json`s used throughout the monorepo
- `eslint-config-zitadel`: ESLint preset

## Login UI

The new implementation of the login is based on Next.js and implements the [OIDC standard](./oidc-standard) by proxying requests to the ZITADEL backend and using the session API.

> The UI shows branding and behaves based on the policies and settings your instance or organization (when using an org scope) has.

Reviewer: ```suggestion
The UI uses the branding and behaves based on the policies and settings of your instance or organization.
```

> It is also able to autenticate users without an OIDC context by navigating to the host of the login, to allow them to self service their account.

Reviewer: ```suggestion
It is also allows users to authenticate directly (without an OIDC context) and self service their account.
```

> It's set of features will increase with time and its progress will be tracked [here](./typescript-repo#implemented-features).

Reviewer: ```suggestion
Its set of features will increase with time and its progress will be tracked [here](./typescript-repo#implemented-features).
```

### Architectural Overview

The new login app consists of a confidential (backend) part and a browser (client side) part.
In addition to that, the OIDC specification builds upon a proxy which is set up in the Next.js middleware. It rewrites requests of the paths `/.well-known/:path*`, `/oauth/:path*`, `/oidc/:path*` to ZITADEL.

> The backend side of the application is authorized using a machine user and a personal access token.

Reviewer: ```suggestion
The backend side of the application is authorized using a machine user's personal access token.
```

The frontend side of the application communicates to the backend side by using a cookie which is set by the server.
The cookie consists of an id and a token and is bound to a session which is updated continuously.

The following illustration shows the architecture of the login and a potential authentication code flow starting from an application which implements the OIDC specification respectively.

![Next.js Login Architecture](/img/typescript-login-architecture.png)

_Note that the illustration is just a representation of the architecture and may not specify actual endpoints._

The flow starts by the `/authorize` endpoint which is intercepted by the login and forwarded to ZITADEL.
ZITADEL interprets the request and specifically redirects back to the login app.
It does so by redirecting to `/login`.
The login is then able to load an [AuthRequest](/docs/apis/resources/oidc_service/oidc-service-get-auth-request#get-oidc-auth-request-details).

The Auth Request defines how users proceed to authenticate. If no special prompts or scopes are set, the login brings up the `/loginname` page.

> The /loginname page allows to enter loginname or email of a user. User discovery is implemented at /api/loginname and if the user is found, redirected to the available authentication method page.

Reviewer: ```suggestion
The /loginname page allows to enter loginname or email of a user. User discovery is implemented at /api/loginname and if the user is found, they will be redirected to the available authentication method page.
```

Right after the user is found, a session is created and set as cookie.
This cookie is then hydrated with more information once the users continues.
The OIDC Auth request is always passed in the url to have a context to the ongoing authentication flow.

> If enough user information is retrieved and the user is authenticated, the flow is finalized by [requesting a the callback url](/docs/apis/resources/oidc_service/oidc-service-create-callback) for the auth request and the user is redirected back to the application.

Reviewer: ```suggestion
If enough user information is retrieved and the user is authenticated accorind to the policies, the flow is finalized by [requesting a the callback url](/docs/apis/resources/oidc_service/oidc-service-create-callback) for the auth request and the user is redirected back to the application.
```

    - [ ] Azure
    - [ ] Apple
- Register

  - [x] Email Password
  - [x] Passkey

#### Self service

The self service part of the application allows to load the login without an OIDC flow.

> After authenticating, or with a valid session, users are able to self service their account.

Reviewer: ```suggestion
Authenticated users are directly able to self service their account.
```

If you want to use the login with your cloud instance, you can purchase your domain on zitadel.com, install it on your domain following [our guide](/docs/guides/manage/cloud/instances#add-custom-domain).
After your domain has been verified, you can reconfigure your DNS settings in order to deploy the login on your own.

#### Deploy to Vercel

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/clone?repository-url=https%3A%2F%2Fgithub.com%2Fzitadel%2Ftypescript&env=ZITADEL_API_URL,ZITADEL_SERVICE_USER_ID,ZITADEL_SERVICE_USER_TOKEN&root-directory=apps/login&envDescription=Setup%20a%20service%20account%20with%20IAM_OWNER%20membership%20on%20your%20instance%20and%20provide%20its%20id%20and%20personal%20access%20token.&project-name=zitadel-login&repository-name=zitadel-login)

To deploy your own version on Vercel, navigate to your instance and create a service user.
Copy its id from the overview and set it as `ZITADEL_SERVICE_USER_ID`.
Then create a personal access token (PAT), copy and set it as `ZITADEL_SERVICE_USER_TOKEN`, then navigate to your instance settings and make sure it gets `IAM_OWNER` permissions.

> Finally set your instance url as `ZITADEL_API_URL`. Make sure to set it without trailing slash.

Reviewer: we might need to add some images

Author: I've added a screenshot
